### DISCUSSION TITLE
render with TextureID and output TextureID

### DISCUSSION THREAD

**Discussion by simonsland** [google/filament#5443] [Q&A]
Recently, I'm trying to use Filament in GPUImageFilter chain, which pass data through textureID.

I wanna know What is the best way for Filament(iOS) to render into a texture created by last GPUImageFilter, and the best way to pass the rendered pixel to the next filter through textureID.

I knew that for Android, it can use GLSurfaceView to render into a texture directly. But for iOS, it seems there is no such way.

**Comment by pixelflinger** [comment]
@bejado should know.

**Reply by simonsland** [reply]
discussion #4406 shows the best way for android is the SurfaceTexture, is there a equivalent way for iOS to solve the problem? @bejado

**Reply by bejado** [reply]
On iOS, you can accomplish something similar by using a `CVPixelBuffer` as the native window. See the [`CONFIG_APPLE_CVPIXELBUFFER` flag](https://github.com/google/filament/blob/059f51338d5999d753324eb868de60b9fa615073/filament/include/filament/SwapChain.h#L168-L179) (note that this is only supported when using Filament's Metal backend). This allows you to render into a `CVPixelBuffer` backed by a texture. Then, you can use either `CVOpenGLESTextureCache` or `CVMetalTextureCache` to get the associated texture (for example, by using `CVOpenGLESTextureCacheCreateTextureFromImage`).

**Reply by simonsland** [reply]
I've tried the way to render into a CVPixelBuffer which have its own pixel data, and the results turns out  that the pixels rendered by filament will cover the input CVPixelBuffer. Is there ways to get the mixed pixel data, or I should mix the input pixel buffer with the filament output by myself?

**Reply by bejado** [reply]
What do you mean by "mixed pixel data"? Do the pixels you render have transparency? You may need to adjust the blend settings in the material.
